admin/cleanup/MergeDuplicateArtistCredits

```diff
@@ -0,0 +1,216 @@
+later version: http://www.gnu.org/licenses/gpl-2.0.txt
+
+=cut
+
+use strict;
+use warnings;
+
+use FindBin;
+use lib "$FindBin::Bin/../../lib";
+
+use Getopt::Long qw( GetOptions );
+use List::MoreUtils qw( natatime );
+use Pod::Usage qw( pod2usage );
+use MusicBrainz::Server::Constants qw(
+    %ENTITIES
+    entities_with
+);
+use MusicBrainz::Server::Context;
+use MusicBrainz::Server::Log qw( log_info );
+
+my $limit = 100;
+my $database = 'MAINTENANCE';
+my $dry_run;
+my $help_flag;
+
+GetOptions(
+    'limit|l=i' => \$limit,
+    'database|d=s' => \$database,
+    'dry-run|n' => \$dry_run,
+    'help|h' => \$help_flag,
+);
+
+pod2usage() if $help_flag;
+pod2usage(
+    -exitval => 64, # EX_USAGE
+    -message => "$0: unrecognized arguments",
+) if @ARGV;
+
+log_info {
+    'Performing a dry run: changes reported below will not be committed ' .
+    'to the database or cache.'
+} if $dry_run;
+
+my $c = MusicBrainz::Server::Context->create_script_context(
+    database => $database,
+);
+
+$c->sql->begin;
+
+my $rows = $c->sql->select_single_column_array(q{
+    WITH ac_names AS (
+        SELECT ac.id,
+               array_agg(
+                row(acn.position,
+                    acn.artist,
+                    acn.name,
+                    acn.join_phrase)
+                ORDER BY acn.position ASC
+               ) AS names
+          FROM artist_credit ac
+          JOIN artist_credit_name acn
+            ON acn.artist_credit = ac.id
+         WHERE ac.edits_pending = 0
+         GROUP BY ac.id
+    )
+    SELECT array_agg(id ORDER BY id ASC) AS duplicates
+      FROM ac_names
+     GROUP BY names
+    HAVING count(id) > 1
+     LIMIT ?
+}, $limit);
+
+unless (@$rows) {
+    log_info { 'No mergeable artist credits found.' };
+    exit(0);
+}
+
+my @ac_entity_models = map { $c->model($_) }
+    sort { $a cmp $b }
+    entities_with('artist_credits', take => 'model');
+
+my $delete_unused_ac_name_query =
+    'DELETE FROM artist_credit_name ' .
+    'WHERE artist_credit = any($1) ' .
+    join(' ', map {
+        'AND NOT EXISTS (' .
+        'SELECT 1 FROM ' . $_->_main_table .
+        ' WHERE artist_credit = any($1)' .
+        ')'
+    } @ac_entity_models) .
+    ' RETURNING artist_credit';
+
+my @cache_entries_to_clear;
+my $updated_entity_count = 0;
+
+for my $duplicates (@$rows) {
+    my ($orig_ac, @dupe_acs) = @$duplicates;
+
+    my $id_display = join(', ', @$duplicates);
+    log_info { "Found duplicate artist credits: $id_display." };
+
+    for my $model (@ac_entity_models) {
+        my $entity_table = $model->_main_table;
+
+        my $entity_ids = $c->sql->select_single_column_array(qq{
+            UPDATE $entity_table SET artist_credit = ?
+             WHERE id IN (SELECT id FROM $entity_table
+                           WHERE artist_credit = any(?::INT[])
+                           ORDER BY id ASC
+                           LIMIT ? FOR UPDATE SKIP LOCKED)
```

Comment:
`FOR UPDATE` will lock the selected entity rows against concurrent updates, and `SKIP LOCKED` will skip ones we can't immediately obtain a lock for (to avoid lock contention). The `SELECT` statements in `$delete_unused_ac_name_query` will still see locked rows, so it won't try to remove the ACs if there are still locked rows left - they'll just have to be processed again in the next hour.